pkg/provisioner/ironic/ironic.go

```diff
@@ -883,9 +884,21 @@ func (p *ironicProvisioner) setCustomDeployUpdateOptsForNode(ironicNode *nodes.N
 		SetTopLevelOpt("deploy_interface", "custom-agent", ironicNode.DeployInterface)
 }
 
+func (p *ironicProvisioner) getInstanceUUID() uuid.UUID {
+	// This procedure generates a predictable UUID that survives re-creating the host
+	return uuid.NewSHA1(uuid.NameSpaceURL, []byte(p.bmcAddress))
+}
+
 func (p *ironicProvisioner) getImageUpdateOptsForNode(ironicNode *nodes.Node, imageData *metal3v1alpha1.Image, bootMode metal3v1alpha1.BootMode, hasCustomDeploy bool, updater *nodeUpdater) {
-	// instance_uuid
-	updater.SetTopLevelOpt("instance_uuid", string(p.objectMeta.UID), ironicNode.InstanceUUID)
+	// Instance UUID marks a node as belonging to this Host.
+	instanceUUID := p.getInstanceUUID().String()
+	// Instance UUID cannot be changed (it serves as a voluntary lock), only deleted and set.
+	if ironicNode.InstanceUUID == "" {
+		updater.SetTopLevelOpt("instance_uuid", instanceUUID, ironicNode.InstanceUUID)
+	} else if ironicNode.InstanceUUID != instanceUUID && ironicNode.InstanceUUID != string(p.objectMeta.UID) {
+		// TODO(dtantsur): make this an error? it may be important for shared Ironic deployments that are shared
```

Comment:
I agree this should be an error. Ironic only returns 409 even though it will never come right, so we do need this detection code.
If we have to recover automatically then we should do it by deleting the Node. I'd prefer we address the root cause in the tests instead, but I'm starting to think that deleting the Node might be safe. Maybe.